Automated Combat and Items Update

varys1337 opened the pull request:
**Update 1 - Opposed rolls pipeline and Combat automation v1**

_The Old Build largely reflected an earlier implementation stage where:_
- Rolls were single-party (unopposed) and resolved immediately.
- Opposed resolution (attacker vs defender) and Degrees of Success were non existant
- Damage and mitigation were presented, but not applied, and armor/resistance handling was manual.
- Hit location on damage cards was determined via a separate random roll, rather than being derived from the attack roll as per RAW.
- Equipment rules (materials/quality, ammunition integration, shield/block nuance) were not encoded, leaving enforcement to players/GM and used older edition's data.

_The New Build introduced:_
- A canonical opposed workflow (combat and skills) driven by targeting and chat cards.
- Centralized, rules-aware TN pipelines with breakdowns.
- Structured handling for DoS/DoF, and configurable bonus damage.
- A real damage engine that computes mitigation and can apply damage to actors.
- Equipment rule encoding for materials, quality tiers, ammunition, and derived “effective” stats - all 3ev4 based
- A world/actor item migration to normalize legacy data and fix schema typos. The items can be set up RAW win the item's attributes section, without the necessity for meriads of "left goblin boot", " right goblin boot" separate stat items.


**1. Detailed Changes**

1.1 **Core Functional Changes (User-visible behavior): Combat Rolls: from immediate single roll → opposed chat-card workflow**

_Old behavior_
Clicking a combat roll on a sheet:
- Opened a basic modifier dialog
- Rolled immediately
- Printed a simple pass/fail chat message

_New behavior_
If a target token is selected:
- Clicking the combat dice creates a pending opposed chat card.
- Attack/defense rolls are triggered from the chat card:
- Attacker: declares attack variant + modifiers
- Defender: selects defense type + modifiers
- Results resolve deterministically and update the same card.
- Unopposed combat style simple checks and just weapon damage rolls are still possible.


1.2 **Skill Rolls: from immediate single roll → unopposed OR opposed (targeted) workflow**

_Old behavior_
- Skill roll = dialog for manual modifier → immediate 1d100 roll → success/failure output.
- Penalties included wounds/fatigue/carry, but there was no explicit TN pipeline with breakdown.

_New behavior_
- No target → normal roll flow with improved TN computation and UI.
- With target(s) → creates opposed skill workflow cards (one per defender target).
- Adds improved option capture (difficulty selection, quick-roll handling, debug support).

1.3 **Chat UI improvements**

_Old behavior_
- Output format was relatively flat and often repeated details in the main body.

_New behavior_
- TN breakdown moved into expandable details <details><summary>TN breakdown</summary>…</details>
- _Subjectively_ cleaner cards
- Debuggable math when needed
- Less clutter for normal play

**2. Rules Implementation Changes** 

**2.1 Hit Location: old random roll → updated RAW-compatible derivation**

_Old build_
Damage cards rolled 1d10 to determine hit location independently of the attack roll.

_Updated build_
- Hit location comes from the ones digit of the attack roll (implemented in combat utilities).
This is a major rules correction because it ties hit location to the actual attack outcome rather than an unrelated random roll.

**2.2 Degrees of Success / Failure (DoS/DoF): absent → fully implemented**

_Old build_
Success/failure was binary; no standardized DoS/DoF pipeline existed.

_Updated build_
Adds a DoS/DoF resolver implementing RAW logic, including:
- Tens digit-based degrees
- TN > 100 rule (tens digit of TN contributes to DoS)
- Lucky/unlucky integration preserved as critical markers
Opposed resolution uses structured results rather than ad hoc comparisons.

**2.3 Damage mitigation + application: manual → computed engine (+ optional auto-apply)**

_Old build_
- Damage rolls existed, but mitigation and HP changes were still mostly manual.
- No consistent model for armor-by-location + resistances + toughness.

_Updated build_
Adds a centralized damage system that supports 8 damage types (physical, fire, frost, shock, poison, magic, silver, sunlight) and computes mitigation via:
- Armor rating (location-aware) ( from the items' update on location based armor)
- Resistances
- Natural Toughness for physical
- Can apply damage automatically (world setting toggle).
- Adds healing support via the same engine.
- Rules alignment improvement
- Mitigation is no longer “table talk math” — it’s system computed, consistently shaped, and repeatable.

**2.4 Block / Shields: previously “just armor-ish” → explicit shield modeling**

_Old build_
- Shields were not modeled as first-class “block-only” mitigation entities; block math was largely left to the table.

_Updated build_
- Adds explicit shield affordances in schema and mitigation:
- Shields do not contribute AR; they contribute Block Rating.
- Block resolution uses shield BR rather than conflating it with armor rating.
- Shield eligibility is enforced for “Block” defense selection.

**2.5 Equipment economics & materials (Chapter 7): previously manual → derived effective stats**

_Old build_
- Weapon/armor/ammo material and quality impacts were not encoded and had older beta edition rules often ( e.g. bodkin arrows)
- Players had to manually bake material effects into damage/enc/price, or ignore them.

_Updated build_
- Adds RAW rule tables to constants and computes derived effective values:
- Weapon quality tiers (inferior/common/superior) → price mult + auto qualities
- Weapon materials (melee vs ranged material tables) → damage mods, encumbrance deltas, enchant levels, price multipliers, auto qualities
- Ammunition materials → damage mods, enchant levels, pricing per 10, derived price-per-shot, auto qualities

These are implemented as derived “effective” fields, avoiding destructive overwrites of stored user inputs. This keeps legacy items compatible while still surfacing RAW math.

**2.6 Ammunition integration: missing/limited → restored ammo selection + combined evaluation**

_Old build_
Ammunition existed as items with a quantity decrement action, but:
- Ranged weapons did not have a robust linkage to ammo choice.
- No unified “weapon + ammo” effective stat model.

_Updated build_
- Adds weapon-side system.ammoId selection (sheet dropdown).
- Adds ammo “effective” calculations and presentation in ammo sheets.
- Supports optional consume behavior (weapon.consumeAmmo) and robust quantity handling ( warnings if no ammunition left and gating the opposed check continuation).

**3. Codebase & Architecture Changes**
**3.1 New module areas introduced**

The updated build adds several new, separable subsystems:

Combat automation modules (module/combat/)
- opposed-workflow.js — canonical multi-step opposed combat workflow using chat cards
- defense-dialog.js — defender dialog: defense type selection + modifiers + TN previews
- tn.js — combat TN pipeline with breakdown objects
- damage-automation.js — mitigation + application engine
- damage-resolver.js — canonical boundary shaping payloads into damage engine calls
- mitigation.js — shield BR, location mapping, mitigation helpers
- chat-handlers.js — centralized chat button handlers for combat/skills workflows

Skills workflow modules (module/skills/)
- skill-tn.js — deterministic skill TN computation with breakdown
- opposed-workflow.js — opposed skill tests via chat cards
- roll-request.js — standardized roll request structures

Migration (module/migrations/)
- items.js — world + embedded item normalization and schema repair

Debug tools (module/dev/)
- Debug settings menu + TN debug macros

**3.2 Modified core files**
- module/entities/actor.js
- Adds defensive initialization of system structures
- Adds wrappers: applyDamage, applyHealing, getDamageReduction
- Adds mobility-derived data for skill TN pipeline use
- Adds aggregation caching for item-derived bonuses
- module/entities/item.js

_Introduces “effective derived stats” for:_

- weapons (damage/enc/price/enchant/qualities)
- ammo (damage/price-per-shot/enchant/qualities)
- Adds structured quality handling and safer non-mutating item-to-actor bonus behaviors

_Sheets + templates (module/sheets/*, templates/*)_
- Adds ammo selection UI
- Adds hit location coverage structures for armor
- Integrates new workflow triggers and breakdown presentation logic

**4. Data Schema & Migration Notes**
**4.1 Template schema updates (template.json)**
- weapon.equipped (fixes legacy typo)
- weapon.qualityLevel, weapon.material, weapon.attackMode
- weapon.ammoId, weapon.consumeAmmo
- weapon.qualitiesStructured, weapon.qualitiesTraits
- ammunition.ammoMaterial, ammunition.arrowType, ammunition.pricePer10
- ammunition.qualitiesStructured, ammunition.qualitiesTraits
- armor.weightClass, armor.effectiveWeightClass, armor.armorClass
- armor.isShield, armor.shieldType
- armor.hitLocations.{Head,Body,Arms,Legs} coverage map

**4.2 Migration: legacy typo repair + defaults**
The new migration explicitly fixes:
- system.equippped → system.equipped (and removes the legacy key)
- Ensures defaults for quality/material/structured arrays across weapons/armor/ammo

Scope
- Migrates world items and embedded actor items
**- Does not automatically migrate compendium packs to operate with new framewrok for items**

**5. Settings & GM Controls**
**5.1 New settings added in updated build**
- autoApplyDamage — automatically apply damage from opposed results
- useDosBonus — add configurable DoS-derived damage bonus
- opposedDebug / opposedDebugFormula / opposedShowResolutionDetails
- skillRollDebug / skillRollLastOptions / skillRollQuickShift
- debugSkillTN — enables TN debug macro tools

These allow production worlds to choose between “automation on” and “chat-only/manual apply” styles.

**6. Notable Technical / Policy Caveats (Important for PR reviewers)**
**6.1 Startup dialog now uses ApplicationV2 constructs ( testing out Appv2 conversion)**

**6.2 Added helper file with deprecated Roll evaluation option (currently unused)**
- module/helpers/opposedRollHelper.js includes Roll.evaluate({ async: true }), which is not the recommended v13 usage (v13 expects awaiting evaluate() without the legacy async option). This file does not appear to be imported anywhere, so it likely does not affect runtime, but reviewers may flag it as dead code / unsafe if later used.

**7. Combat workflow with opposed checks quick-report**
- Target selected → clicking combat roll creates pending opposed chat card (no auto-roll).
- “Roll Attack” opens attacker dialog; roll posts DSN-compatible Roll message.
- Defender can select Evade/Parry/Block/Counter; Block disabled if no shield.
- TN breakdown expands correctly and shows circumstance/manual/variant entries only there.
- Resolution updates the original chat card deterministically.
- If autoApplyDamage enabled, defender HP updates and wound thresholds update.

Damage and mitigation
 - Physical damage uses location AR + natToughness.
-  Elemental/magic damage respects resistances.
- Shields do not contribute to AR; block uses BR.
- “Apply Damage” buttons work from chat cards.

Equipment and ammo
- Weapon sheet shows ammunition selector; selecting ammo persists on weapon.
- Ammo item shows derived effective stats (damage/pricing) correctly.
- Ammo consumption reduces quantity without going negative and warns at 0.
- Material/quality changes reflect in effective stats without overwriting base fields.

Migration
-  Legacy items with equippped migrate to equipped.
-  No console errors on world load; migration logs (if any) are clean.

**8. File Change Summary** 
module/entities/actor.js, module/entities/item.js
module/handlers/init.js, module/handlers/startup.js, module/entrypoint.js
module/helpers/skillCalcHelper.js

Sheets: 
module/sheets/actor-sheet.js, npc-sheet.js, merchant-sheet.js, item-sheet.js
Templates: templates/* (weapon/ammo/armor/item/npc/actor)
module/constants.js (massively expanded into rule tables)
template.json (expanded item schema)

New files:
Combat automation suite: module/combat/*
Skills pipeline: module/skills/*
Migration: module/migrations/items.js
Debug utilities: module/dev/*

varys1337 committed:
Refactor _onClickCharacteristic and _onProfessionsRoll

varys1337 committed:
Refactor npc-sheet for damage automation

varys1337 committed:
Refactor _onSpellRoll for safety and readability

Refactor _onSpellRoll and related methods for improved safety and readability. Implement optional chaining and nullish coalescing for better handling of actor data.

varys1337 committed:
Clean up _onSpellRoll and remove _onResetResource

Removed unused comments and the _onResetResource function.

varys1337 committed:
Add files via upload

Copilot committed:
Initial plan

Copilot committed:
guard: module/entities/actor.js — defensive system access + aggregator

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

Copilot committed:
guard: module/entities/item.js — defensive container handling

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

Copilot committed:
guard: module/sheets/merchant-sheet.js — defensive system access and safe price handling

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

Copilot committed:
guard: module/sheets/npc-sheet.js — defensive system access

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

Copilot committed:
guard: templates — handlebars guards and numeric defaults

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

varys1337 committed:
Merge pull request #10 from varys1337/copilot/fix-guarded-system-accesses-one-more-time

Guard nested system property access and add single-pass item aggregation

Copilot committed:
Initial plan

Copilot committed:
guard: defensive system access in sheet files (merchant, actor, item)

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

Copilot committed:
guard: templates — handlebars guards and numeric defaults for containers

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

Copilot committed:
guard: merchant-sheet spell and resistance dialogs — safe characteristic access

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

varys1337 committed:
Merge pull request #11 from varys1337/copilot/fix-guarded-system-accesses-please-work

Guard nested .system property access and fix duplicate function

Copilot committed:
Initial plan

Copilot committed:
guard: module/sheets/actor-sheet.js — defensive container handling

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

Copilot committed:
guard: module/sheets/npc-sheet.js — defensive system access

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

varys1337 committed:
Merge pull request #12 from varys1337/copilot/fix-guarded-system-accesses-08e2391d-e4f3-40a1-a3b2-31193e569b53

Complete defensive system access guards for sheet files

varys1337 committed:
Refactor speed, initiative, and wound calculations

Copilot committed:
Initial plan

Copilot committed:
guard: module/entities/actor.js — defensive system access + aggregator

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

Copilot committed:
guard: module/entities/actor.js — defensive system access + aggregator (complete)

Copilot committed:
guard: module/entities/item.js — defensive container handling

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

Copilot committed:
guard: module/sheets/merchant-sheet.js — defensive system access and safe price handling

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

Copilot committed:
guard: module/combat — defensive system reads for combat helpers

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

Copilot committed:
Initial plan

varys1337 committed:
Merge pull request #13 from varys1337/copilot/fix-guarded-system-accesses-523e61f9-c90b-4853-af8f-0d8a7228c036

Defensive guards for nested .system property access and single-pass item aggregator

Copilot committed:
guard: module/entities/actor.js — defensive system access + aggregator

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

Copilot committed:
Fix: Use isEquipped consistently for armorEnc calculation

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

varys1337 committed:
Merge branch 'main' into copilot/fix-guarded-system-accesses-cc87ef4f-fda8-4d8b-803a-2a007e125e85

varys1337 committed:
Merge pull request #14 from varys1337/copilot/fix-guarded-system-accesses-cc87ef4f-fda8-4d8b-803a-2a007e125e85

Guard nested system access and add single-pass item aggregator

Copilot committed:
Initial plan

Copilot committed:
guard: module/entities/item.js — defensive container handling

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

varys1337 committed:
Merge pull request #15 from varys1337/copilot/fix-item-container-safety

Defensive container handling in module/entities/item.js

varys1337 committed:
Refactor _calculateENC and related methods

varys1337 committed:
Refactor container item preparation and price modification

varys1337 committed:
Refactor price modification methods for clarity

varys1337 committed:
Update item.js

varys1337 committed:
Refactor item deletion logic in actor-sheet.js

varys1337 committed:
Implement CombatRoll class for combat mechanics

varys1337 committed:
Implement AdvantageDialog class for combat options

varys1337 committed:
Refactor prepareData and add system data initialization

Refactor prepareData method and add _ensureSystemData method to initialize required system data structures and ensure safe defaults.

varys1337 committed:
Refactor getData() in actor-sheet.js

Refactor getData() method to improve structure and add safe defaults.

varys1337 committed:
Implement CombatHelper for opposed combat rolls

varys1337 committed:
Refactor combat resolution methods in combatHelper

varys1337 committed:
Update npc-sheet.js

varys1337 committed:
Add opposed roll resolution helper function

varys1337 committed:
Update _onDamageRoll to use opposed roll initiation

Refactor damage roll handler to initiate opposed roll.

varys1337 committed:
Create damageHelper.js

varys1337 committed:
Refactor damage calculations and armor handling

varys1337 committed:
Refactor opposed roll functions and add new features

Refactor opposed roll logic to improve clarity and functionality. Added executeOpposedRoll and displayOpposedRollResult functions.

varys1337 committed:
Refactor damage roll and apply damage logic

Refactor damage roll logic to include combat helper for applying damage and improved hit location determination. Update async handling for weapon rolls and damage application.

varys1337 committed:
Delete documents directory

varys1337 committed:
Create Chapter 1 - Getting Started

varys1337 committed:
Add files via upload

varys1337 committed:
Refactor activateListeners for improved clarity

varys1337 committed:
Enhance damage reduction calculations with hit location

varys1337 committed:
Include hitLocation in damage calculation

Added hitLocation parameter to damage calculation.

varys1337 committed:
Implement DefenseDialog for defense selection

varys1337 committed:
Refactor damage roll logic and improve readability

varys1337 committed:
Update npc-sheet.js

varys1337 committed:
Update chat-handlers.js

Copilot committed:
Initial plan

Copilot committed:
Fix all JavaScript syntax errors in sheet files

Co-authored-by: varys1337 <170896219+varys1337@users.noreply.github.com>

varys1337 committed:
Merge pull request #16 from varys1337/copilot/fix-js-syntax-errors

Fix JavaScript syntax errors from spaces before dot operators

varys1337 committed:
Remove apply damage button click handler

Removed the apply damage button click handler from the NPC sheet.

varys1337 committed:
Update damage-automation.js

varys1337 committed:
Refactor damage automation to eliminate redundancy

Removed duplicate total calculation in damage automation.

varys1337 committed:
Refactor damage roll logic and error notifications

Fix damage roll functionality and improve error handling.

varys1337 committed:
Refactor status tag handling for clarity

varys1337 committed:
Fix damage roll binding in actor-sheet.js

varys1337 committed:
Clean up comments in damage-automation.js

Removed unnecessary comments from the damage automation module.

varys1337 committed:
Fix formatting in template.json for armor attributes

Added armor location parts

varys1337 committed:
Fix formatting and add hit locations section

Added armor locations, replaced the olde item-selection-container

varys1337 committed:
Refactor damage automation for clarity and consistency

Refactor damage automation code for consistency and clarity, including adjustments to variable names and formatting.

varys1337 committed:
Refactor hitLocations keys to camelCase format

varys1337 committed:
Fix hit locations checkbox names and formatting

varys1337 committed:
Fix formatting of checkbox names in armor-sheet.html

varys1337 committed:
Fix HP update logic and chat message styles

Refactor damage and healing application logic to update HP correctly for linked and unlinked tokens. Fix chat message type to use styles instead of types.

varys1337 committed:
Enhance unconscious effect application logic

Improved the application of the unconscious effect by checking if it already exists on the target actor before applying it. Added error handling for the effect application process.

varys1337 committed:
Create diceHelper.js

varys1337 committed:
Refactor hit location logic to use attack roll result

Replaced rollHitLocation function with getHitLocationFromRoll that determines hit location based on the ones digit of the attack roll result.

varys1337 committed:
Refactor toughness calculation in damage automation

Updated toughness calculation to use natToughness from actorData.

varys1337 committed:
Update actor-sheet.js

varys1337 committed:
Update actor-sheet.js

varys1337 committed:
Refactor hit location logic and damage table

Refactor hit location calculation and update damage table structure.

varys1337 committed:
Refactor contentString generation for clarity

varys1337 committed:
Update actor-sheet.js

varys1337 committed:
Implement hit location and attack success calculations

Added hit location calculation and degrees of success for attack rolls.

varys1337 committed:
Update actor-sheet.js

varys1337 committed:
Refactor combat roll and damage logic

Refactor combat roll and damage handling logic for improved clarity and functionality.

varys1337 committed:
Update parsing, correct syntax errors

varys1337 committed:
Implement rollHitLocation for backward compatibility

Added a backwards-compatible helper function to derive hit location from an attack roll result or roll 1d100 as a fallback.

varys1337 committed:
Implement defense choice prompt in attack helper

Added defense type prompting for attacker/GM workflow.

varys1337 committed:
Refactor damage automation for clarity and accuracy

varys1337 committed:
Refactor skill value retrieval in DefenseDialog

varys1337 committed:
Refactor hit location and damage calculation logic

varys1337 committed:
Enhance damage button with additional data attributes

varys1337 committed:
Enhance damage application with dosBonus and penetration

Added support for dosBonus and penetration parameters in damage application.

varys1337 committed:
Refactor damage automation for clarity and consistency

Updated damage automation system to enhance damage type handling, improve code readability, and ensure consistent parameter usage across functions.

varys1337 committed:
Add UUIDs for attacker and defender in opposed rolls

varys1337 committed:
Enhance getHitLocationFromRoll for input validation

Updated getHitLocationFromRoll to handle non-numeric input safely.

varys1337 committed:
Improve defense skill handling and damage roll logic

Refactor defense skill calculation and update damage roll logic.

varys1337 committed:
Normalize rank values and update roll evaluations

Added rank value normalization for legacy typos and updated roll evaluation calls to remove async flag.

varys1337 committed:
Refactor DefenseDialog for improved functionality

varys1337 committed:
Remove rank selection dropdown from actor sheet

varys1337 committed:
Update Lore_Roll_b2K1CAkdbH7vutmq.yml

varys1337 committed:
Refactor actor property access in Ride macro

varys1337 committed:
Refactor Subterfuge roll logic for clarity

varys1337 committed:
Refactor actor system property access

Refactored property access for actor system and commerce values to use direct access instead of defensive checks.

varys1337 committed:
Refactor damage roll evaluation and cleanup UUIDs

varys1337 committed:
Enhance chat handlers for combat automation

Updated chat handlers for combat automation to be compatible with Foundry VTT v13. Added support for damage types and improved event handling for Apply Damage and Apply Healing actions.

varys1337 committed:
Fix roll evaluation for Foundry V13 changes

Updated Roll#evaluate to reflect changes in Foundry V13.

varys1337 committed:
Refactor combat roll handling to use OpposedWorkflow

varys1337 committed:
Refactor chat handlers for UESRPG 3ev4

varys1337 committed:
Add files via upload

varys1337 committed:
Refactor chat handlers for v13 compatibility

varys1337 committed:
Update roll handling and message card flags

varys1337 committed:
Refactor DefenseDialog to improve defense option handling

varys1337 committed:
Enhance opposed workflow with improved logic

Refactor opposed workflow to improve TN derivation and defense handling.

varys1337 committed:
Delete copilot-fix-guarded-system-accesses.patch

varys1337 committed:
Update chat-handlers.js

varys1337 committed:
Refactor defense dialog for improved functionality

Refactor defense selection dialog to improve functionality and clarity. Updated combat style handling, added manual modifier input, and improved TN calculations for various defense options.

varys1337 committed:
Refactor opposed workflow for clarity and functionality

Refactor opposed workflow to improve clarity and functionality, including updates to combat styles, actor resolution, and rendering logic.

varys1337 committed:
Refactor opposed test handling in actor-sheet.js

varys1337 committed:
Add styles for defender dialog button layout

Added styles for the defender dialog buttons to match the attack dialog.

varys1337 committed:
Refactor defense dialog buttons and layout

varys1337 committed:
Refactor dialog handling in opposed workflow

varys1337 committed:
Add opposed debug logging setting to game config

varys1337 committed:
Add files via upload

varys1337 committed:
Refactor combat style handling and add debug logging

Refactor combat workflow to centralize combat style logic and improve debugging features.

varys1337 committed:
Refactor DefenseDialog for TN calculations

Refactor defense dialog to centralize TN calculations and improve code organization.

varys1337 committed:
Add files via upload

varys1337 committed:
Refactor damage roll calculation in attack-helper.js

Updated damage roll calculation to use weapon's getDamageRoll method for better automation.

varys1337 committed:
Add skill opposed action handling to chat handlers

varys1337 committed:
Improve damage type determination from weapon qualities

Refactor getDamageTypeFromWeapon to handle weapon qualities more robustly, including structured qualities and a fallback to legacy qualities.

varys1337 committed:
Implement armor mobility penalties and refactor calculations

Added armor mobility penalties and refactored related calculations for effective armor weight class. Updated item handling to ensure proper penalties are applied based on equipped armor.

varys1337 committed:
Implement dice bonus function and enhance item prep

Added a function to apply a flat bonus to dice strings and updated item preparation methods to include new quality handling.

varys1337 committed:
Register new game settings for skill rolls

Added several new game settings for skill roll diagnostics and UI enhancements.

varys1337 committed:
Add formatDegree function to roll object

varys1337 committed:
Add files via upload

varys1337 committed:
Refactor skill roll logic and add mobility penalties

Refactor skill roll logic to include mobility penalties and improve handling of skill item retrieval.

varys1337 committed:
Enhance item sheet with new options and calculations

Added new item options and improved data processing for armor and ammunition types.

varys1337 committed:
Update constants.js

varys1337 committed:
Update entrypoint.js

varys1337 committed:
Add Skill TN Debug utility for GM inspection

This module provides a developer utility for inspecting Skill TN computation in GM mode. It allows debugging of skill calculations through a global helper function.

varys1337 committed:
Implement attack context module for combat items

This module provides a canonical derivation layer for classifying combat-facing items. It includes a function to derive the attack/defense context from a given item.

varys1337 committed:
Add skill opposed workflow functionality

Implement skill opposed workflow for combat scenarios, allowing attackers and defenders to roll skills with various options and track outcomes.

varys1337 committed:
Implement skill roll request module

This module implements the internal data contract for skill roll workflows, including functions for debugging, clamping values, normalizing options, building requests, and validating requests.

varys1337 committed:
Implement skill TN computation for UESRPG 3ev4

This module implements skill TN computation for UESRPG 3ev4, providing a breakdown of skill difficulties and modifiers based on various factors such as actor state and item bonuses.

varys1337 committed:
Add item migration and normalization script

This migration script normalizes items in the game, ensuring correct properties for weapons, armor, and ammunition. It includes functions to infer attack modes and handle legacy data.

varys1337 committed:
Enhance styles for structured qualities and item sheets

Added styles for structured qualities and improved layout for item sheets.

varys1337 committed:
Enhance ammunition sheet with effective hints and options

varys1337 committed:
Enhance armor sheet with new fields and options

varys1337 committed:
Change input types to number for various fields

varys1337 committed:
Update template.js to accomodate item and opposed rolls update

varys1337 committed:
Rename COMBAT_AUTOMATION.md to Combat-Automation.md

varys1337 committed:
Rename SKILL_TN_PIPELINE.md to Features: Skill-Automation .md

varys1337 committed:
Reduce dialog width and update defense options

varys1337 committed:
Refactor variant labels and defense handling

varys1337 committed:
Refactor combat style handling for NPCs

varys1337 committed:
Add files via upload

varys1337 committed:
Add actor select debug registration for GM

varys1337 committed:
Refactor roll logic and enhance dialog UI

Updated the roll formula to use a constant and improved the dialog UI for roll options.

varys1337 committed:
Enhance professions roll with permission check

Refactor professions roll handling to include permission check and improve functionality.

varys1337 committed:
Refactor skill and profession listing functions

varys1337 committed:
Update roll-request.js

varys1337 committed:
Refactor item skill bonus collection logic

Refactor skill item bonus collection logic to use normalized keys and improve deduplication.

varys1337 committed:
Add SYSTEM_ROLL_FORMULA constant

Added central roll formula constant for tests.

varys1337 committed:
Update template.json

varys1337 committed:
Update defense dialog styles for consistency

varys1337 committed:
Add optional parameters for damage application

varys1337 committed:
Change fallback roll from 1d100 to 1d10

varys1337 committed:
Implement armor coverage normalization in damage automation

Added armor coverage normalization logic to handle legacy data and improve damage calculation accuracy.

varys1337 committed:
Updates , too many

varys1337 committed:
Refactor dice expression normalization and item stats

Refactor damage and bonus calculations to use normalizeDiceExpression for consistency. Update armor and weapon preparation methods to derive effective stats based on quality and material profiles.

varys1337 committed:
Use SYSTEM_ROLL_FORMULA in doTestRoll function

varys1337 committed:
Refactor weapon damage and hit location logic

Refactor weapon damage calculation and hit location logic. Introduce structured qualities and improve damage roll handling.

varys1337 committed:
Add armor and shield options to item sheet

Added options for armor materials and classes, and shield types.

varys1337 committed:
Refactor damage handling and roll evaluations

Refactor damage handling in NPC sheet to centralize chat-card button handling and avoid duplicate registrations. Update hit location and damage roll evaluations for improved clarity and functionality.

varys1337 committed:
Define armor and shield qualities and materials

Added armor and shield quality definitions, including materials and profiles.

varys1337 committed:
Enhance CSS for weapon damage chat cards

varys1337 committed:
Update armor and magic attributes with effective values

varys1337 committed:
Update input fields for effective values in ammunition sheet

varys1337 committed:
Update armor-sheet.html

varys1337 committed:
Update NPC sheet to use effective armor values

varys1337 committed:
Refactor weapon-sheet to handle effective values

varys1337 committed:
Add armor and shield properties to template.json

varys1337 committed:
Delete documents/Features: Combat-Automation .md

varys1337 committed:
Delete documents/Features: Skill-Automation .md

varys1337 committed:
Refactor damage application to use applyDamageResolved

varys1337 committed:
Enhance armor coverage logic in damage automation

Refactor damage automation logic for armor and shields, improving coverage determination and scoring for forceful impacts.

varys1337 committed:
Add files via upload

varys1337 committed:
Update defense-dialog.js

varys1337 committed:
Refactor damage application and button attributes

varys1337 committed:
Add files via upload

varys1337 committed:
Enhance action points and weapon selection logic

Refactor action points handling and improve weapon selection logic. Added combat circumstance modifiers in the attacker declaration dialog.

varys1337 committed:
Introduce circumstanceMod for combat calculations

Added circumstanceMod parameter to combat TN calculations.

varys1337 committed:
Add files via upload

varys1337 committed:
Add newline at end of item.js

Fix missing newline at end of file.

varys1337 committed:
Add debugging settings and update config options

Added new debugging settings and updated existing ones to disable configuration in the UI.

varys1337 committed:
Refactor applyDamage to use damage resolver

Refactored applyDamage method to use damage resolver for applying damage, improving maintainability and compatibility.

varys1337 committed:
Improve target selection handling in actor-sheet.js

Handle multiple target selections by notifying the user and using the first target.

varys1337 committed:
Add newline at end of template.json

Fix missing newline at end of file

varys1337 committed:
Delete documents directory

varys1337 committed:
Add files via upload

varys1337 committed:
Refactor startup dialog implementation

Refactor startup dialog to use Dialog instead of ApplicationV2. Update changelog rendering and cleanup event listeners.

BokuNC commented:
Testing this and found a problem. The text editor on the character sheets and items, for some reason, lack the typeable fields. You can still use it by editing it in "show code mode". 

Is this system using TinyMce? Maybe something changed. Didn't check to see if i could identify the cause, but i'll take a look later.

varys1337 commented:
> Testing this and found a problem. The text editor on the character sheets and items, for some reason, lack the typeable fields. You can still use it by editing it in "show code mode".
> 
> Is this system using TinyMce? Maybe something changed. Didn't check to see if i could identify the cause, but i'll take a look later.

Could you be more specific about char sheets and items you refer to? Some items' header values inputs were locked off on purpose. The new material,quality,etc. framework adapts those values automatically. e.g. inferior-partial-iron should set the effective price values, AR / BR and weight category in accordance with 3ev4 equipment tables.

TinyMCE is part of core foundry sheets afaik, so no problems should occur, I guess. I have not seen any changes, but I picked this system development only on v13.351

varys1337 committed:
Add files via upload

varys1337 committed:
Implement deterministic Active Effect modifier evaluation

This file implements deterministic Active Effect modifier evaluation for roll-time pipelines, including functions for handling numeric coercion, typed damage detection, effect priority determination, and collecting effect sources from actors.

varys1337 committed:
Add files via upload

varys1337 committed:
Update chat-handlers.js

varys1337 committed:
Update combat-utils.js with chat handlers

Refactor combat utility functions and add chat handlers for UESRPG 3ev4.

varys1337 committed:
Modify Natural Toughness to affect all damage types

Updated Natural Toughness to apply to all damage types instead of just physical. Adjusted damage calculation logic to reflect changes in toughness and final damage output.

varys1337 committed:
Refactor damage resolver functions and imports

varys1337 committed:
Implement ammo pre-consumption for ranged attacks

Added ammunition pre-consumption logic for ranged attacks and updated weapon damage roll function to handle pre-consumed ammo.

varys1337 committed:
Add combat TN modifier functions and logic

varys1337 committed:
Add files via upload

varys1337 committed:
Implement Active Effects and resource management

Add support for Active Effects modifiers and resource management in SimpleActor.

varys1337 committed:
Implement hooks for Active Effect handling

varys1337 committed:
Enhance startup dialog functionality and structure

Refactor startup dialog handling for compatibility and UI improvements.

varys1337 committed:
Refactor biography enrichment and effects handling

Refactor biography enrichment and add active effects handling.

varys1337 committed:
Refactor SimpleItemSheet for improved readability

Refactor SimpleItemSheet methods and improve code consistency.

varys1337 committed:
Implement Active Effects control in NPC sheet

Added functionality to handle Active Effects in the Effects tab.

varys1337 committed:
Enhance skill TN computation with active effects

Refactor skill TN computation to include active effects and combat TN modifiers.

varys1337 committed:
Implement AE keys dump functionality for GM

Added a helper function to dump AE keys to the console for GM use.

varys1337 committed:
Add styles for Active Effects in main.css

Added styles for Active Effects section in AppV1 sheets.

varys1337 committed:
Add effects tab to actor sheet template

varys1337 committed:
Add debug settings form for GM and client

varys1337 committed:
Add effects tab to ammunition sheet

varys1337 committed:
Add effects tab to armor sheet template

varys1337 committed:
Implement effects tab in combat style sheet

Added effects tab with controls for managing effects.

varys1337 committed:
Add effects tab to container sheet template

varys1337 committed:
Add effects tab to faction sheet template

varys1337 committed:
Add effects tab to item sheet template

varys1337 committed:
Add effects tab to language sheet template

varys1337 committed:
Implement effects tab in magic skill sheet

Added effects tab with controls for managing effects.

varys1337 committed:
Add effects tab to NPC sheet template

varys1337 committed:
Add effects tab to power sheet template

varys1337 committed:
Implement effects tab in skill sheet

Added effects tab with controls for managing effects.

varys1337 committed:
Refactor skill sheet HTML for better structure

varys1337 committed:
Refactor skill sheet layout and navigation

Updated skill sheet layout by replacing inputs with labels and select elements for better user experience. Removed unused sections and adjusted the tab navigation.

varys1337 committed:
Implement effects tab in spell sheet

Added effects tab with controls for managing effects.

varys1337 committed:
Add effects tab to talent sheet template

varys1337 committed:
Add effects tab to trait sheet template

varys1337 committed:
Add effects tab to weapon sheet template

varys1337 committed:
Delete uesrpg-system-attribute-keys.md

varys1337 committed:
Add combat modifiers to template.json

BokuNC commented:
> > Testing this and found a problem. The text editor on the character sheets and items, for some reason, lack the typeable fields. You can still use it by editing it in "show code mode".
> > Is this system using TinyMce? Maybe something changed. Didn't check to see if i could identify the cause, but i'll take a look later.
> 
> Could you be more specific about char sheets and items you refer to? Some items' header values inputs were locked off on purpose. The new material,quality,etc. framework adapts those values automatically. e.g. inferior-partial-iron should set the effective price values, AR / BR and weight category in accordance with 3ev4 equipment tables.
> 
> TinyMCE is part of core foundry sheets afaik, so no problems should occur, I guess. I have not seen any changes, but I picked this system development only on v13.351

Sure can do.

Both on items and Character Sheets we have major TinyMce fields;

On the Character Sheet its the space for Biography, once opened for editing, it doesn't allow you to type into it directly.
<img width="488" height="253" alt="image" src="https://github.com/user-attachments/assets/59e44b7a-f944-44ad-a8e4-bdff085b37b7" />

It also happen in the Description field for the Items overall

<img width="568" height="232" alt="image" src="https://github.com/user-attachments/assets/d5340579-0c9c-4495-8e38-cbafd75beda5" />

On the Basic "Item" sheet, the add Modifier button isn't doing anything either. (ok, this does seem to be a problem on the master branch as well)

<img width="221" height="365" alt="image" src="https://github.com/user-attachments/assets/7940aa48-9206-49c3-a34b-45ff2cf533cb" />

Hopefully its an easy fix for the text editor. Their containers just need a larger min-height.
On uesrpg-3ev4>styles>main.css; Line 1669, changing min-height from 0 to 100 solves the issue.

Starting on line 1665, just in case:
"/* 3) ProseMirror/editor containers should not demand full height unless parent is constrained */
.worldbuilding .editor,
.worldbuilding .prosemirror {
  height: auto !important;
  min-height: 100 !important;
}"

Thats all i've changed.

BokuNC commented:
Okay, applied your new commits and we have a weird bug. It doesn't seem to be loading sheets.
<img width="973" height="281" alt="image" src="https://github.com/user-attachments/assets/9722a0f9-ca5f-4014-b1f2-9578c9339743" />
Players, NPCs or Items, that's all we have when trying to create a new or open a old one. Did create a new world just to check.

<img width="505" height="532" alt="image" src="https://github.com/user-attachments/assets/cfdaa35e-aeb7-4c59-92bc-cfcd405236dd" />

varys1337 commented:
> Okay, applied your new commits and we have a weird bug. It doesn't seem to be loading sheets. <img alt="image" width="973" height="281" src="https://private-user-images.githubusercontent.com/119072605/530993409-9722a0f9-ca5f-4014-b1f2-9578c9339743.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjcxMDM2MjIsIm5iZiI6MTc2NzEwMzMyMiwicGF0aCI6Ii8xMTkwNzI2MDUvNTMwOTkzNDA5LTk3MjJhMGY5LWNhNWYtNDAxNC1iMWYyLTk1NzhjOTMzOTc0My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIzMFQxNDAyMDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xMDE0ZTM4MWY4OWE0OTIwMjZlNjYzN2FkOWMwZTc1ZjU4NmIzY2E5NTNiNWJhZDU5NDhiOTE4ZTE1YWM0N2FiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Jnj8ekND0aPIu_0mGwVsQRTQqW8EwOmLj1LP7wWsQEQ"> Players, NPCs or Items, that's all we have when trying to create a new or open a old one. Did create a new world just to check.
> 
> <img alt="image" width="505" height="532" src="https://private-user-images.githubusercontent.com/119072605/530993705-cfdaa35e-aeb7-4c59-92bc-cfcd405236dd.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjcxMDM2MjIsIm5iZiI6MTc2NzEwMzMyMiwicGF0aCI6Ii8xMTkwNzI2MDUvNTMwOTkzNzA1LWNmZGFhMzVlLWFlYjctNGM1OS05MmJjLWNmY2Q0MDUyMzZkZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMjMwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTIzMFQxNDAyMDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04ZThkYWMzYWRmMzM0MWI2MWFlNzM4YTRiMDgyZWViYWFhNzYwOTA4NTY5YTk2YzkzNzAxMGU2NTQ1OGFjM2IwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9._7lxdFZSQp376AgzYPDGGW2Q_1uwuYsNtcP7V1quNbI">

I am mid update and forgot to turn off continuous changes added into PR - don't run the until you see an Update message report.

Thank you for the previous report about text editor, will be fixed in the next update.
